### DISCUSSION TITLE
Trying to understand ifstmt.getThen() = block

### DISCUSSION THREAD

**Discussion by CaledoniaProject** [github/codeql#7731] [General]
I'm reading this doc https://codeql.github.com/docs/codeql-language-guides/basic-query-for-java-code/

And the query is:

```
import java

from IfStmt ifstmt, Block block
where ifstmt.getThen() = block and
  block.getNumStmt() = 0
select ifstmt, "This 'if' statement is redundant."
```

Why can't I use this instead? The block variable looks redundant to me.

```
import java

from IfStmt ifstmt, Block block
where ifstmt.getThen().getNumStmt() = 0
select ifstmt, "This 'if' statement is redundant."
```

**Comment by smowton** [comment]
You can. I imagine the doc example was just trying to show how multiple variables can be declared in the `from` line. In the latter example you can remove `block` leaving

```
from IfStmt ifstmt
where ifstmt.getThen().(BlockStmt).getNumStmt() = 0 // Type name Block is deprecated, but is equivalent to BlockStmt
select ifstmt, "This 'if' statement is redundant."
```

**Comment by CaledoniaProject** [comment]
@smowton IfStmt.getThen() returns a Stmt variable, and it's a different type of variable to Block, how could they be equal?

**Comment by CaledoniaProject** [comment]
Looks like both BlockStmt and IfStmt inherits Stmt class, therefore I can associate the two variables.
Because codeql does not support type conversion like `BlockStmt(ifstmt.getThen()).getNumStmt()`, I will have to use two variables to do the job.

Am I understanding it correctly?

**Reply by smowton** [reply]
Doh right, you actually need `where ifstmt.getThen().(BlockStmt).getNumStmt() = 0`. I've updated my response above. That `.(type)` syntax is the CodeQL syntax for an typecast.

Note that in CodeQL a cast basically consists of a check that the thing being cast matches the characteristic predicate of the target type, returning no result if it doesn't. Classes can also overlap, and there is no requirement for the thing being cast to have "originally" had any particular type, unlike typical object-oriented languages where in order to successfully cast to type T, the object in question has to have been constructed with at least that type -- for example, I can have:

```
class OneTwo extends int { OneTwo() { this = [1, 2] } }
class TwoThree extends int { TwoThree() { this = [2, 3] } }

select 2.(OneTwo).(TwoThree) // result = 2
select 1.(OneTwo) // result = 1
select 1.(TwoThree) // no result
select 3.(TwoThree) // result = 3
select 3.(OneTwo) // no result
```

The numbers on the LHS weren't "constructed" with any particular type, they're just ordinary integers, but I can cast them to `OneTwo` or `TwoThree` which is equivalent to constraining them to match `this = [1, 2]` and `this = [2, 3]` respectively.

In the case above where we cast a `Stmt` to a `BlockStmt`, we check that it in fact is a `BlockStmt` (`BlockStmt` doesn't have a characteristic predicate, but it does extend the database type `@block`, so the cast checks that the database entity really is a `@block`), and it also makes the methods of that class, including `getNumStmt()`, visible.

**Reply by CaledoniaProject** [reply]
You have a typo in the post: the constructor of class TwoThree should be TwoThree() instead :-)

**Reply by smowton** [reply]
Fixed :)

**Comment by CaledoniaProject** [comment]
@smowton I actually tried to look up `@block` today, but I only find this one:

```
java/ql/lib/semmle/code/java/Statement.qll:class BlockStmt extends Stmt, @block {
```

So where is it defined?

**Reply by smowton** [reply]
If you ctrl-click / go-to-defintion, it should take you to the dbscheme file. All types beginning with `@` are database types: primitives that are provided by the extractor.

**Comment by CaledoniaProject** [comment]
Running more tests, and I have another weird behavior with this query:

```
import java

from IfStmt ifstmt, BlockStmt block
where ifstmt.getThen().(BlockStmt).getNumStmt() = 0
select ifstmt, block, "This 'if' statement is redundant."
```

Problems

1. The `block` variable is "undefined" but codeql still consider the query as valid
2. Having the `block` variable in the select statement causes duplicated results, was it a bug?

**Reply by smowton** [reply]
Because `block` is unconstrained, you will get a result for every pair of  an `IfStmt` matching the query and any `BlockStmt` in the whole program. You could see a similar cartesian product behaviour for example with `from int x, int y where x = [1, 2, 3] and y = [1, 2, 3] select x, y` -- you'll get all pairs of solutions for the unrelated variables x and y -- `{1, 1}, {1, 2}, {1, 3}, {2, 1}, ...`

If you want to report a block, keep the old version of the query. If you don't, remove both `BlockStmt block` from the `from` line and ` block` from the select.

This is only valid QL btw because all database types are fundamentally finite, so if you don't constrain them at all you get all of them. If you tried the same thing with an `int` instead of a `BlockStmt` which ultimately derives from a database type then QL would refuse to evaluate a query with an infinite result.
